pkg/server/server.go

```diff
@@ -541,5 +552,21 @@ func (s *ProxyServer) serveRecvBackend(stream agent.AgentService_ConnectServer,
 			klog.Warningf("<<< Unrecognized packet %+v", pkt)
 		}
 	}
+
 	klog.Infof("<<< Close backend %v of agent %s", stream, agentID)
+
+	frontends, _ := s.getFrontends(agentID)
```

Comment:
This closes a huge hole which is awesome. I believe however there is still a much smaller hole in pending dial. Most of the time I would expect pending dial to be empty. However if there is something in there, there is a chance its request went out via this backend. If so we will never get the response and that also needs to be dealt with. 

The issue is that we do not record in the pending data structure which backend it used, so we cannot tell if anything on the pending list would be effected by a given backend breaking. We also need to work out how to deal with it. One option would be to just fail, which is probably easiest. However as the connection has not yet be established, we should be able to switch to using a different backend.

Reply:
The worst case scenario for PendingDial is for the client to hang for a bit and hit the 30s dial timeout that we specify in our konnectivity client. I agree this is a much smaller issue since it won't cause any indefinite hanging behavior, and I think we should address this in a separate PR. 

Created https://github.com/kubernetes-sigs/apiserver-network-proxy/issues/126

Reply:
Might be good to a TODO as a placeholder with the bug id.

Reply:
Good idea, added.

Reply:
Agree that the 30s dial timeout makes pending dial less detrimental.